driver/dma_control.c

```diff
@@ -83,38 +83,54 @@ struct dma_control {
 /* DMA. */
 
 
-static void reset_dma_controller(struct dma_control *dma)
+static int reset_dma_controller(struct dma_control *dma)
 {
     writel(CDMACR_Reset, &dma->regs->cdmacr);
 
     /* In principle we should wait for the reset to complete, though it doesn't
      * actually seem to take an observable time normally.  We use a deadline
      * just in case something goes wrong so we don't deadlock. */
-    unsigned long deadline = jiffies + 2;
+    unsigned long deadline = jiffies + msecs_to_jiffies(1);
     int count = 0;
-    while (readl(&dma->regs->cdmacr) & CDMACR_Reset  &&
+    int reset_stat = 0;
+    while ((reset_stat=readl(&dma->regs->cdmacr) & CDMACR_Reset)  &&
            time_before(jiffies, deadline))
         count += 1;
 
+    if (reset_stat)
+        return -EIO;
+
     /* Now restore the default working state. */
-    writel(0x00008000, &dma->regs->sa_msb);
     writel(CDMACR_IrqEn | CDMACR_Err_IrqEn, &dma->regs->cdmacr);
+    return 0;
 }
 
 
-static void maybe_reset_dma(struct dma_control *dma)
+static bool check_dma_status(struct dma_control *dma)
```

Comment:
I'm in two minds about `bool` vs `int` for error codes.  The problem is, we want to be consistent ... and most functions seem to be returning an error code.  Not sure it's possible to be fully consistent, to be honest.  Maybe the fact that we're logging an error is a hint that this should be an error code?

Reply:
Done